tl_codegen/src/generator.rs

```diff
@@ -0,0 +1,568 @@
+                            assert!(data.bindings.is_empty());
+
+                            split.extend(data.types.into_iter().flat_map(|ty| {
+                                match ty {
+                                    syn::Ty::Path(None, path) => split_path_to_global_segments(namespaces, path),
+                                    _ => unreachable!(),
+                                }
+                            }));
+                        },
+                        syn::PathParameters::Parenthesized(_) => unreachable!(),
+                    }
+                }
+
+                split
+            };
+
+            match ty.to_type_ir().unwrap().unboxed() {  // FIXME
+                syn::Ty::Path(None, path) => split_path_to_global_segments(namespaces, path),
+                syn::Ty::Tup(tys) => {
+                    tys.into_iter().flat_map(|syn_ty| {
+                        match syn_ty {
+                            syn::Ty::Path(None, path) => split_path_to_global_segments(namespaces, path),
+                            ty => unreachable!("`syn::Ty` encountered: {:#?}", ty),
+                        }
+                    }).collect()
+                },
+                ty => unreachable!("`syn::Ty` encountered: {:#?}", ty),
+            }
+        }
+
+        fn get_input_data<'a>(namespaces: &[String], ctor: &'a Constructor)
+            -> ConstructorInputData<'a>
+        {
+            let unrolled_output_syn_ty = get_unrolled_syn_ty(namespaces, &ctor.output);
+            let kind = match ctor.output.to_type_ir().unwrap().kind { // FIXME
+                TypeIrKind::Dynamic => TypeckKind::Dynamic,
+                _ => {
+                    let is_dynamic = ctor.fields.iter()
+                        .any(|f| f.ty.to_type_ir().unwrap().kind == TypeIrKind::Dynamic); // FIXME
+
+                    if is_dynamic { TypeckKind::Dynamic } else { TypeckKind::Static }
+                },
+            };
+            let deps = ctor.fields.iter().map(|f| get_unrolled_syn_ty(namespaces, &f.ty)).collect();
+
+            ConstructorInputData {
+                ty: unrolled_output_syn_ty,
+                kind: kind,
+                deps: deps,
+                ctor: ctor,
+            }
+        }
+
+        let mut all_ctors_input_data = vec![];
+
+        for (namespaces, ty_data) in &self.types {
+            for (_ty_name, ty_ctors) in ty_data {
+                for ty_ctor in &ty_ctors.0 {
+                    let input_data = get_input_data(&namespaces, ty_ctor);
+                    all_ctors_input_data.push(input_data);
+                }
+            }
+        }
+
+        for (namespaces, fn_ctors) in &self.functions {
+            for fn_ctor in fn_ctors {
+                let input_data = get_input_data(&namespaces, fn_ctor);
+                all_ctors_input_data.push(input_data);
+            }
+        }
+
+        all_ctors_input_data
+    }
+}
```

Comment:
well this sure deletes all the quoting. :(
I find it a lot less readable to construct `syn` structs instead of using `quote!`

Reply:
In code I write, I'd rather trade readability and conciseness for robustness and predictability.  _That said_, I'd be **super** happy to know if there is something like `syn!` macro which is akin to `quote!` but returns type-safe `syn` structures. This would combine the best of both worlds.

Reply:
This works:
```rust
syn::parse_item(quote! {
    pub const LAYER: i32 = #layer;
}.as_str()).unwrap();
```
which is essentially the wanted `syn!` macro albeit with a cost of runtime parsing.

I'm going to replace some really obnoxious cases when pure-`syn` code can take up to 50 lines but `syn::parse_*` + `quote! {}.as_str()` one - only 5-10.